src/arch/x86/x86_64/src/amdromehsmp.rs

```diff
@@ -0,0 +1,131 @@
+#![allow(non_snake_case)]
+
+use crate::pci::config32;
+use crate::pci::PciAddress;
+use vcell::VolatileCell;
+
+const HSMP_MESSAGE_ID_SMN: u32 = 0x3B1_0534;
+const HSMP_MESSAGE_RESPONSE_SMN: u32 = 0x3B1_0980;
+const HSMP_MESSAGE_ARGUMENT_0_SMN: u32 = 0x3B1_09E0;
+const HSMP_MESSAGE_ARGUMENT_1_SMN: u32 = 0x3B1_09E4;
+const HSMP_MESSAGE_ARGUMENT_2_SMN: u32 = 0x3B1_09E8;
+const HSMP_MESSAGE_ARGUMENT_3_SMN: u32 = 0x3B1_09EC;
+const HSMP_MESSAGE_ARGUMENT_4_SMN: u32 = 0x3B1_09F0;
+const HSMP_MESSAGE_ARGUMENT_5_SMN: u32 = 0x3B1_09F4;
+const HSMP_MESSAGE_ARGUMENT_6_SMN: u32 = 0x3B1_09F8;
+const HSMP_MESSAGE_ARGUMENT_7_SMN: u32 = 0x3B1_09FC;
+
+pub struct SmuServiceRequest {
+    command: u32,
+    data: [u32; 8],
+}
+
+#[derive(Debug)]
+pub struct SmuServiceResponse {
+    status: u32,
+    data: [u32; 8],
+}
+
+pub struct HSMP<'a> {
+    NB_SMN_INDEX_3: &'a VolatileCell<u32>,
+    NB_SMN_DATA_3: &'a VolatileCell<u32>,
+}
+
+// trait: pub unsafe fn test(&self, v: u32) -> Result<u32, u32>
+
+impl HSMP<'_> {
+    pub fn new(nbio: usize) -> Self {
+        let bus = match nbio {
+            0 => 0,
+            1 => 0x20,
+            2 => 0x40,
+            3 => 0x60,
+            _ => panic!("invalid NBIO"),
```

Comment:
Hmm, have to read up some more. I only have one socket populated, so not sure.

Also, hardcoding those PCI bus addresses only works because Ron's to-be-merged patch sets them up that way.